### PR TITLE
Add support for the client permission change

### PR DESCRIPTION
**Pull request by kishikawakatsumi**
Related: https://github.com/realm/realm-sync-services/pull/236

Add support for the client permission change.
This PR add:
- Data schema for the client permission (`PermissionChange` and a base class)
- Add convenience method for retrieving `__management.realm`
- Unit tests

The permission management requires the same table in both Realm and RealmSwift. I created `PermissionChange` class in both Realm and RealmSwift. (Then add `@compatibility_alias RLMSyncPermissionChange` for the Cocoa naming conventions. But maybe it is unnecessary.)

Unit tests will fail due to the same class exists in Realm and RealmSwift.

```
realm-cocoa/Realm/ObjectServerTests/SwiftSyncTestCase.swift:55: error: -[ObjectServerTests.SwiftObjectServerTests testBasicSwiftSync] : failed: caught "RLMException", "RLMObject subclasses with the same name cannot be included twice in the same target. Please make sure 'PermissionOffer' is only linked once to your current target."
```

Do you have any ideas avoiding it? @mrackwitz @austinzheng @jpsim 

For unit tests, it needs to add the following configurations to `configuration.yml` before tests.

```
auth:
  sync_hosts:
    - 'localhost:9080'
```

How should I achieve it? @austinzheng 

CC @mrackwitz @cmelchior @bigfish24

### CODE REVIEW

**Comment by kishikawakatsumi** [comment]
Failing `testPermissionChange**GrantNoChanges` tests is a known issue. It will be resolved   by https://github.com/realm/realm-sync-services/pull/474

File: RealmSwift/Sync.swift [hunk 1]
```diff
@@ -153,6 +155,69 @@ extension SyncUser {
...
+
+        if let realmUrl = realm?.configuration.syncConfiguration?.realmURL.absoluteString {
+            self.realmUrl = realmUrl
+        }
+
+        if let userId = user?.identity {
+            self.userId = userId
+        }
+        
+        self.mayRead.value = mayRead;
```
Comment (mrackwitz) [inline RealmSwift/Sync.swift]: Semicolons here and below. 👮

File: RealmSwift/Sync.swift [hunk 2]
```diff
@@ -235,6 +300,69 @@ extension SyncUser {
...
+
+        if let realmUrl = realm?.configuration.syncConfiguration?.realmURL.absoluteString {
+            self.realmUrl = realmUrl
+        }
+
+        if let userId = user?.identity {
+            self.userId = userId
+        }
+
+        self.mayRead.value = mayRead;
```
Comment (mrackwitz) [inline RealmSwift/Sync.swift]: Semicolons here and below. 👮

File: Realm/ObjectServerTests/RLMObjectServerTests.m [hunk 3]
```diff
@@ -582,4 +582,865 @@ - (void)testLogBackInReopenRealmDownload {
     }
 }
 
+- (void)testPermissionChangeForRealmAndUserGrantAdministrativeAccess {
```
Comment (bmunkholm) [inline Realm/ObjectServerTests/RLMObjectServerTests.m]: From a cursory look, it seems there is a lot of copy/paste code in these tests. Can this be refactored?
Also adding a single code comment before each test section would improve readability.

**Comment by jpsim** [comment]
I'm hoping this can be superseded by #4324.

**Comment by jpsim** [comment]
superseded by #4324.

**Pull request closed without merging**